File: client/src/main/java/org/camposmdev/client/FourSoulsApp.java
package org.camposmdev.client;

import com.almasb.fxgl.app.ApplicationMode;
import com.almasb.fxgl.app.CursorInfo;
import com.almasb.fxgl.app.GameApplication;
import com.almasb.fxgl.app.GameSettings;
import javafx.scene.input.KeyCode;
import javafx.scene.paint.Color;
import org.camposmdev.client.entity.factory.FSEntityFactory;
import org.camposmdev.client.model.Game;
import org.camposmdev.client.service.BoardPosition;
import org.camposmdev.client.service.EntityService;
import org.camposmdev.client.ui.scene.FSSceneFactory;
import org.camposmdev.client.ui.view.ActionDrawerView;
import org.camposmdev.client.ui.view.PlayMatView;

import java.util.Map;

import static com.almasb.fxgl.dsl.FXGL.*;

public class FourSoulsApp extends GameApplication {
    public static void main(String[] args) {
        launch(args);
    }
    private static final int APP_WIDTH = 1600, APP_HEIGHT = 900;
    @Override
    protected void initSettings(GameSettings options) {
        options.setApplicationMode(ApplicationMode.DEVELOPER);
        options.setTitle("Four Souls");
        options.setVersion("1.0.0-alpha");
        options.setAppIcon("icons/soul_circle.png");
        options.setWidth(APP_WIDTH);
        options.setHeight(APP_HEIGHT);
        options.setPreserveResizeRatio(true);
        options.setScaleAffectedOnResize(true);
        if (options.getApplicationMode() == ApplicationMode.RELEASE) {
            options.setIntroEnabled(true);
            options.setMainMenuEnabled(true);
        }
        options.setGameMenuEnabled(true);
        options.setSceneFactory(new FSSceneFactory());
        options.setManualResizeEnabled(true);
        options.setFullScreenAllowed(true);
        options.setPauseMusicWhenMinimized(true);
        options.setDefaultCursor(new CursorInfo("cursor.png", 0, 0));
        options.setFontGame("EdmundMcMillen_v2.ttf");
        options.setFontUI("EdmundMcMillen_v2.ttf");
        options.setFontText("EdmundMcMillen_v2.ttf");
        options.getCSSList().add("main.css");
        options.addEngineService(EntityService.class);
    }

    @Override
    protected void onPreInit() {
        final double VOLUME = 0.1;
        getSettings().setGlobalMusicVolume(VOLUME);
        getSettings().setGlobalSoundVolume(VOLUME);
        getNotificationService().setBackgroundColor(Color.web("#2D2D30"));
        getNotificationService().setTextColor(Color.WHITE);
        /* load texture and store in cache */
        getAssetLoader().loadTexture("board.jpg");
    }

    @Override
    protected void initUI() {
        ActionDrawerView.instance().render();
    }

    @Override
    protected void initGameVars(Map<String, Object> vars) {
        /* if the player chose singleplayer, create a game var */
        var game = Game.create();
        game.shuffle();
        vars.put("game", game);
        /* otherwise send request to server */
    }

    @Override
    protected void initGame() {
        /* start the music */
//        loopBGM("The Binding of Isaac - 11 Repentant.mp3");
        /* add the four souls entity factory to spawn the game entities */
        getGameWorld().addEntityFactory(new FSEntityFactory());
        var es = getService(EntityService.class);
        /* render in the play mat */
        var playMat = new PlayMatView(-500);
        playMat.render();
        /* add treasure deck to game world */
        var treasureEntity = getGameWorld().spawn("treasure_deck");
        es.mapper().set(treasureEntity, BoardPosition.CENTER_LEFT);
        /* add loot deck to game world */
        var lootEntity = getGameWorld().spawn("loot_deck");
        es.mapper().set(lootEntity, BoardPosition.CENTER);
        /* add monster deck to game world */
        var monsterEntity = getGameWorld().spawn("monster_deck");
        es.mapper().set(monsterEntity, BoardPosition.CENTER_RIGHT);
        /* add player to game world */
        es.spawnPlayer("b-isaac");
    }

    @Override
    protected void initInput() {
        /* set key listener to fullscreen game */
        onKeyDown(KeyCode.F11, () -> {
            var fullScreen = getSettings().getFullScreen().get();
            getSettings().getFullScreen().set(!fullScreen);
        });
    }
}

File: client-app/src/main/java/org/camposmdev/client/FourSoulsApp.kt
package org.camposmdev.client

import com.almasb.fxgl.app.ApplicationMode
import com.almasb.fxgl.app.CursorInfo
import com.almasb.fxgl.app.GameApplication
import com.almasb.fxgl.app.GameSettings
import com.almasb.fxgl.dsl.*
import javafx.scene.input.KeyCode
import javafx.scene.paint.Color
import org.camposmdev.client.entity.factory.FSEntityFactory
import org.camposmdev.client.service.BoardPosition
import org.camposmdev.client.service.EntityService
import org.camposmdev.client.ui.scene.FSSceneFactory
import org.camposmdev.client.ui.view.ActionDrawerView
import org.camposmdev.client.ui.view.PlayMatView
import org.camposmdev.model.atlas.MasterCardAtlas
import org.camposmdev.model.game.LocalGameManager
import org.camposmdev.util.FXUtil

class FourSoulsApp : GameApplication() {
    companion object {
        private const val TITLE = "Four Souls"
        private const val VERSION = "1.0.0-alpha"
        private const val APP_WIDTH = 1600
        private const val APP_HEIGHT = 900

        @JvmStatic
        fun main(args: Array<String>) {
            launch(FourSoulsApp::class.java, args)
        }
    }

    override fun initSettings(settings: GameSettings) {
        settings.title = TITLE
        settings.version = VERSION
        settings.width = APP_WIDTH
        settings.height = APP_HEIGHT
        settings.appIcon = "icons/soul_circle.png"
        settings.isPreserveResizeRatio = true
        settings.isScaleAffectedOnResize = true
        settings.isIntroEnabled = false
        settings.isMainMenuEnabled = false
        settings.isGameMenuEnabled = true
        settings.sceneFactory = FSSceneFactory()
        settings.isManualResizeEnabled = true
        settings.isFullScreenAllowed = true
        settings.isPauseMusicWhenMinimized = true
        settings.defaultCursor = CursorInfo("cursor.png", 0.0, 0.0)
        settings.fontGame = "EdmundMcMillen_v2.ttf"
        settings.fontUI = "EdmundMcMillen_v2.ttf"
        settings.fontText = "EdmundMcMillen_v2.ttf"
        settings.cssList += "main.css"
        settings.applicationMode = ApplicationMode.DEVELOPER
        settings.addEngineService(EntityService::class.java)
    }

    override fun onPreInit() {
        val volume = 0.1
        getSettings().globalMusicVolume = volume
        getSettings().globalSoundVolume = volume
        getNotificationService().backgroundColor = Color.web("#2D2D30")
        getNotificationService().textColor = Color.WHITE
        /* load texture and store in cache */
        getAssetLoader().loadTexture("board.jpg")
    }

    override fun initUI() {
        ActionDrawerView.instance().render()
    }

    override fun initGameVars(vars: MutableMap<String, Any>) {
        val atlas = FXUtil.loadJSON("cards.json", MasterCardAtlas::class.java)
        val game = LocalGameManager(atlas)
        /* shuffle the cards */
        game.shuffle()
        /* add the game to game variables*/
        vars["game"] = game
        /* otherwise send request to server */
    }

    override fun initGame() {
        /* start the music */
        // loopBGM("The Binding of Isaac - 11 Repentant.mp3");
        /* add the four souls entity factory to spawn the game entities */
        getGameWorld().addEntityFactory(FSEntityFactory())
        val es = EntityService.get()
        /* render in the play mat (background) */
        val playMat = PlayMatView(-500)
        playMat.render()
        /* add treasure deck to game world */
        val treasureDeck = es.spawn().treasure_deck()
        es.mapper().position(treasureDeck, BoardPosition.CENTER_LEFT)
        /* add loot deck to game world */
        val lootDeck = es.spawn().loot_deck()
        es.mapper().position(lootDeck, BoardPosition.CENTER)
        /* add monster deck to game world */
        val monsterDeck = es.spawn().monster_deck()
        es.mapper().position(monsterDeck, BoardPosition.CENTER_RIGHT)
        /* add player to game world */
        val characterId = "b-isaac"
        es.spawn().player(characterId)
        //        ItemShopView itemShop = new ItemShopView();
//        itemShop.render();
//        var m1 = es.spawn_monster();
//        var c = m1.getComponent(MonsterCardComponent.class);
//        var root = (TilePane) itemShop.content();
//        root.setTranslateX(getAppWidth()/4d - root.getPrefWidth()/2d);
//        root.setTranslateY(getAppHeight()/2d - root.getPrefHeight()/2d);
//        root.getChildren().add(c.texture());
    }

    override fun initInput() {
        /* set key listener to fullscreen game */
        onKeyDown(KeyCode.F11) {
            val fullScreen: Boolean = getSettings().fullScreen.get()
            getSettings().fullScreen.set(!fullScreen)
        }
    }
}


